tracer/src/tracer.rs

```diff
@@ -140,20 +138,17 @@ impl TraceBlock {
 
 impl Drop for TraceBlock {
     fn drop(&mut self) {
-        // SAFETY: start has been initialised as part of initialising the value of TRACER
-        let start = unsafe { TRACER.get().unwrap().start };
+        let start = TRACER.get().unwrap().start;
         let trace_event = TraceEvent {
             timestamp: self.start.duration_since(start),
             event: self.event,
             end_timestamp: Some(Instant::now().duration_since(start)),
-            // SAFETY: thread_depth() returns a number local to the current thread
-            depth: unsafe { TRACER.get().unwrap().thread_depth() },
+            depth: TRACER.get().unwrap().thread_depth(),
         };
-        // SAFETY: add_event and decrease_thread_depth access current thread only specific data
-        unsafe {
-            TRACER.get_mut().unwrap().add_event(trace_event);
-            TRACER.get_mut().unwrap().decrease_thread_depth();
-        }
```

Comment:
I gather the original implementation here is trying to reduce the locking overhead as much as possible, but it seems we couldn't guarantee that these two steps here wouldn't get intervened, maybe we should consider using `Arc<RwLock<Tracer>>` as the type of `TRACER` that kind of thing 🧐

Reply:
And drop `AtomicU64` and `Mutex` inside `Tracer` since it's behind a `RwLock`